scylla-cdc-printer/src/printer.rs

```diff
@@ -0,0 +1,294 @@
+                    self.set_upper_timestamp(timestamp).await;
+                }
+            }
+
+            let now = SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .unwrap();
+            if now >= self.end_timestamp.to_std()? {
+                return Ok(());
+            }
+        }
+    }
+
+    async fn set_upper_timestamp(&self, new_upper_timestamp: Duration) {
+        for reader in self.readers.iter() {
+            reader.set_upper_timestamp(new_upper_timestamp).await;
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use scylla::batch::Consistency;
+    use scylla::query::Query;
+    use scylla::SessionBuilder;
+    use tokio::time::sleep;
+
+    use super::*;
+
+    const SECOND_IN_MILLIS: i64 = 1_000;
+    const TEST_KEYSPACE: &str = "test";
+    const TEST_TABLE: &str = "t";
+    const SLEEP_INTERVAL: i64 = SECOND_IN_MILLIS / 10;
+    const WINDOW_SIZE: i64 = SECOND_IN_MILLIS / 10 * 3;
+    const SAFETY_INTERVAL: i64 = SECOND_IN_MILLIS / 10;
+
+    fn get_create_table_query() -> String {
+        format!("CREATE TABLE IF NOT EXISTS {}.{} (pk int, t int, v text, s text, PRIMARY KEY (pk, t)) WITH cdc = {{'enabled':true}};",
+                TEST_KEYSPACE,
+                TEST_TABLE
+        )
+    }
+
+    async fn create_test_db(session: &Arc<Session>) -> anyhow::Result<()> {
+        let mut create_keyspace_query = Query::new(format!(
+            "CREATE KEYSPACE IF NOT EXISTS {} WITH REPLICATION = {{'class': 'SimpleStrategy', 'replication_factor': 1}};",
+            TEST_KEYSPACE
+        ));
+        create_keyspace_query.set_consistency(Consistency::All);
+
+        session.query(create_keyspace_query, &[]).await?;
+        session.await_schema_agreement().await?;
+
+        // Create test table
+        let create_table_query = get_create_table_query();
+        session.query(create_table_query, &[]).await?;
+        session.await_schema_agreement().await?;
+
+        let table_name_under_keyspace = format!("{}.{}", TEST_KEYSPACE, TEST_TABLE);
+        session
+            .query(format!("TRUNCATE {};", table_name_under_keyspace), &[])
+            .await?;
+        session
+            .query(
+                format!("TRUNCATE {}_scylla_cdc_log;", table_name_under_keyspace),
+                &[],
+            )
+            .await?;
+        Ok(())
+    }
+
+    async fn populate_db_with_pk(session: &Arc<Session>, pk: u32) -> anyhow::Result<()> {
+        let table_name_under_keyspace = format!("{}.{}", TEST_KEYSPACE, TEST_TABLE);
+        for i in 0..3 {
+            session
+                .query(
+                    format!(
+                        "INSERT INTO {} (pk, t, v, s) VALUES ({}, {}, 'val{}', 'static{}');",
+                        table_name_under_keyspace, pk, i, i, i
+                    ),
+                    &[],
+                )
+                .await?;
+        }
+        session.await_schema_agreement().await?;
```

Comment:
Waiting for schema agreement is only necessary after you make schema changes (create/modify a keyspace/table). You aren't doing any schema changes in this function.